### PR TITLE
Feature: add refund checkbox to the framework

### LINKED ISSUE #6657: Gateway API should add a refund checkbox automatically on the donation admin screen
## User Story

Currently, when changing a donation status on the admin screen to "refund", the checkbox to refund it on the gateway side is not automatically displayed by the Framework. In fact, is necessary that each gateway implementation add the checkbox in the admin screen, which is not ideal - the ideal is to have this refund checkbox added automatically.

## Acceptance Criteria

-   [ ] The refund checkbox is displayed automatically on new gateway implementations

### PR DESCRIPTION
**Pull request by glaubersilva**
Resolves #6657

## Pre-review Checklist

-   [ ] Acceptance criteria satisfied and marked in related issue
-   [ ] Relevant `@unreleased` tags included in DocBlocks
-   [ ] Includes unit tests
-   [ ] Reviewed by the designer (if follows a design)
-   [ ] [Self Review](https://give.gitbook.io/development-manual/devops/github/code-reviews#self-review) of code and UX completed

### CODE REVIEW

**Comment by glaubersilva** [comment]
@jonwaldstein I liked the button idea, seems to be more intuitive from the UX perspective. About the `handleRefundDonation` I suppose that it would be used like this: `$registeredGateway->handleRefundDonation($donation);` instead of calling the `refundDonation` method directly [as I did here](https://github.com/impress-org/givewp/pull/6658/files#diff-f04cd0d24b8efa092be5aef9c1f4dabb3f58a25084f84ce8289b850fc30228c6R239), right?

**Comment by jonwaldstein** [comment]
@glaubersilva yup exactly!  I think for now we can leave the ux the way it is since this page is going to be redesigned in the near future but we'll definitely revisit when that time comes.  Here's what i'm thinking to wrap this one up:

1. Let's add a method `handleRefundDonation` to `PaymentGateway` (similar to createPayment) with a try / catch.  The catch will handle any Exceptions.  Basically, If a refund did not work the gateway can just throw a `PaymentGatewayException` and our framework will deliver the message.
2. This method will handle the `PaymentRefunded` command (I just realized this exists) using the `PaymentRefundedHandler` (also exists lol).  This really just ensures the framework is setting the refunded status on the donation and redirecting to the appropriate location.

**Comment by glaubersilva** [comment]
@jonwaldstein I did implement your suggestions and change the PR description to better explain how it works after the changes.

File: src/Framework/PaymentGateways/PaymentGateway.php [hunk 1]
```diff
@@ -497,4 +547,17 @@ private function isFunctionImplementedInGatewayClass(string $methodName): bool
...
+     * Checks to see if the provided method is returning a specific command type as result.
+     * It always returns 'false' if the provided method doesn't explicitly declare the return type.
+     *
+     * @unreleased
+     */
+    private function isFunctionReturningCommand(string $methodName, PaymentCommand $command): bool
+    {
+        $reflector = new ReflectionMethod($this, $methodName);
+
+        return $reflector->getReturnType() ? $reflector->getReturnType()->getName() : '' === get_class($command);
```
Comment (jonwaldstein) [inline src/Framework/PaymentGateways/PaymentGateway.php:561]: @glaubersilva looks like we can't use `$reflector->getReturnType()->getName()` until php 7.1 (we support 7.0 right now).  My other comment might end up removing this anyway.

Screenshot 2023-02-23 at 4 26 06 PM

File: src/Framework/PaymentGateways/PaymentGateway.php [hunk 2]
```diff
@@ -76,6 +78,21 @@ public function __construct(SubscriptionModule $subscriptionModule = null)
...
+     *             by default, it returns 'false' to prevent:
+     *             1) Duplication of refund checkboxes on gateways that already implement it;
+     *             2) Adding the refund checkbox on the admin screen in gateways with an empty refundDonation() method;
+     *             3) Overwriting legacy refund methods with empty refundDonation() methods.
+     *             4) Break gateways that implement the refundDonation() method but don't return the PaymentRefunded command as result
+     */
+    public function supportsRefund(): bool
+    {
+        return $this->isFunctionReturningCommand('refundDonation', (new PaymentRefunded()));
+    }
```
Comment (jonwaldstein) [inline src/Framework/PaymentGateways/PaymentGateway.php:94]: @glaubersilva thanks for really thinking through this! I love how you outlined all the scenarios - it's really helpful for me to understand what's going on.

I do wonder if we're being a little too strict on the implementation of a return type though, as some folks might be building gateways and not be used to doing that sort of thing in php. 

For that reason, (if possible) i'm thinking we should just add a `refundDonation()` to the main PaymentGateway class that just returns false and let `supportsRefund()` use `$this->isFunctionImplementedInGatewayClass('refundDonation')`.  I like the simplicity of saying that a gateway supports refunds if it's using the refundDonation method.

To prevent duplication maybe we can use some kind of combination of `doing_filter` and/or `doing_action` before proceeding in the legacy adapter.  Another options is in our legacy adapter in the js script we just check for the checkbox in the dom to make sure it doesnt' exist yet.

Open to your thoughts here.

**Review by jonwaldstein — CHANGES_REQUESTED** [review]

**Comment by JasonTheAdams** [comment]
@jonwaldstein @glaubersilva I would like to prioritize getting this resolved during cooldown, please.

File: src/Framework/PaymentGateways/PaymentGateway.php (hunk 2 above)
Comment (glaubersilva) [reply src/Framework/PaymentGateways/PaymentGateway.php]: Thanks for the insights! I tried an approach with the `doing_action` but it seems not to be the most suitable for this situation, so I implemented a JS approach here: 1e486288916efdb506bee0d57655c257a44ad4d2

Basically what the changes implemented in the commit mentioned above are doing is: 

1. Ensure that the framework is the last one to register a callback in the `give_view_donation_details_totals_after` hook.
2. Before rendering the checkbox on the hook from the previous step, verify if there is just one hidden checkbox in the DOM's proper place, if there is more than one, means that the gateway already added it, so the framework will do nothing.

**Comment by glaubersilva** [comment]
@jonwaldstein Thanks for the insights! I did implement your suggestions and left a comment explaining some details. I also changed the PR description again. Let me know if there is anything more to improve. =)

File: src/LegacyPaymentGateways/Adapters/LegacyPaymentGatewayAdapter.php [hunk 3]
```diff
@@ -178,4 +179,73 @@ private function getOrCreateDonor(
...
+            </div>
+            <script>
+                if (!!document.getElementById('give-payment-status') &&
+                    1 === document.querySelectorAll('div.give-admin-box > div.give-hidden input[type="checkbox"]').length
+                ) {
+                    document.getElementById('give-payment-status').addEventListener('change', function (event) {
+                        const refundCheckbox = document.getElementById('give-gateway-opt-refund');
+
+                        if (null === refundCheckbox) {
+                            return;
```
Comment (jonwaldstein) [inline src/LegacyPaymentGateways/Adapters/LegacyPaymentGatewayAdapter.php:210]: @glaubersilva wouldn't we need to hide this checkbox if one already exists?  We might want to be more specific about targeting the checkbox also like checking if the id attribute has the string `opt-refund` in it to make sure it's the right checkbox.

Comment (glaubersilva) [reply src/LegacyPaymentGateways/Adapters/LegacyPaymentGatewayAdapter.php]: Resolved in 1c9da987e4d771d2006e0489352deffc221ed8fb

About your question:

> wouldn't we need to hide this checkbox if one already exists?

That's not necessary because each checkbox adds its own script to handle the visibility and the framework default checkbox already starts hidden because it uses the give-hidden class.

File: src/Framework/PaymentGateways/PaymentGateway.php [hunk 4]
```diff
@@ -123,6 +134,32 @@ public function handleCreatePayment(Donation $donation)
...
     }
 
+    /**
+     * @unreleased
+     */
+    public function handleRefundDonation(Donation $donation)
+    {
+        try {
+            $command = $this->refundDonation($donation);
+            $this->handleGatewayPaymentCommand($command, $donation);
```
[the remark this answers is not in the document]
Comment (glaubersilva) [reply src/Framework/PaymentGateways/PaymentGateway.php]: Resolved in c01e3140d80abece9fc8e34eb9f7d11ca963a086

**Comment by glaubersilva** [comment]
@jonwaldstein I finally moved the existing code to the controller class, so now there is no more merge conflict. Let me know if I can merge this one or if there is anything more to adjust. Thanks! =)

**Review by jonwaldstein — APPROVED** [review]
@glaubersilva great job man, this looks good to me.  please send this to QA to test on a few different gateways.

**Comment by jonwaldstein** [comment]
@glaubersilva one more thing, would you mind adding a test to `GatewayPaymentControllerTest` for the new refund method? You should be able to duplicate what's already there for the create method and swap a few things out to return the refunded command.

**Comment by JasonTheAdams** [comment]
Ignore. Just testing Slack.

**Comment by JasonTheAdams** [comment]
Testing again.

**Comment by Benunc** [comment]
This has been tested and is good to merge. Will need to run all automated tests still. But it passes manual tests both with existing gateways and with the example gateway repo.

**Review by Benunc — APPROVED** [review]
